[FSDP2] Moved managed states to `device`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* __->__ #117861
* #117852
* #117817
* #117814
* #117776
* #117984
* #118186
* #117994

This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[FSDP2] Moved managed states to `device`

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Moved managed states to `device`"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Moved managed states to `device`"


This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Moved managed states to `device`"


This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -83,3 +84,31 @@ def _get_managed_states(
+    """
+    We have FSDP move states to device for simpler and faster initialization
+    since FSDP almost always uses CUDA for training. We move parameters/buffers
+    rather than modules to allow ignoring specific states in the future.
```
Nit: I still spend some time to understand the differernce between ``module.to(device)`` and ``param/buffer.to(device)``. 

maybe a even more straightforward docstring

```We move parameters/buffers rather than modules since modules may contain ignored parameters/buffers```

Commit:
Update on "[FSDP2] Moved managed states to `device`"


This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented:
I envision that this `fully_shard` `device` arg can take CUDA, CPU, or meta. For this discussions, we mostly care about CUDA and meta. The semantics are that the sharded parameters are on the specified `device` by when the `fully_shard` call returns. If `device="meta"`, then we need a separate API to materialize, and this API will probably need to call some FSDP-specific functions so that FSDP gets the correct references to the newly materialized parameters.

---

> 3. Apply TP and FSDP (forgot which order we picked)

I propose that TP must be applied before FSDP from a runtime perspective. More precisely, I mean that FSDP must shard the tensor already sharded by TP. FSDP's module-level all-gather always runs before TP's op-level collectives, so FSDP's all-gather must undo the FSDP sharding before a TP op runs.

@wanchaol mentioned that there are some complications for `(Shard(0), Shard(0))` placement if we shard with TP then FSDP from the state dict perspective. We should flesh this issue out to get everyone on the same page.

---

> In (3), will FSDP know how to initialize the model, if it came 'meta initialized'? Or is the assumption that FSDP would 'empty init' on device, and a step (4) would be done explicitly?

If FSDP is always applied as the last parallelism, then the current meta-device initialization already works. FSDP sees a meta-device module, runs some initialization (either user-defined via a passed-in function or via `reset_parameters()` defined on the module), and then shards the module. If the meta-device module already has DTensors, we expect DTensor's `__torch_dispatch__` to handle the initialization ops correctly.

The main upside I could see for having an explicit step (4) is that the user may not use FSDP, e.g. the user could just use PP + TP. In that case, we still need a meta-device init solution. In that case, I do not think that it should affect FSDP's `device` arg, or at least, I do not see how it would block us from having `fully_shard(device="cuda")` semantics as in this PR.

Commit:
Update on "[FSDP2] Moved managed states to `device`"


This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -81,3 +82,32 @@ def _get_managed_states(
+        if tensor.device == device:
+            continue
+        if isinstance(tensor, DTensor):
+            if (dtensor_mesh_type := tensor._spec.mesh.device_type) != device.type:
```
nit: let's try to stick with public API if we can `tensor.device_mesh.device_type`

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -81,3 +82,32 @@ def _get_managed_states(
+                    "Requires DTensor to have mesh of the same type as the FSDP mesh "
+                    f"but got {dtensor_mesh_type} for DTensor and {device.type} for FSDP"
+                )
+            raise AssertionError(
```
this assertion error seems always happening if it's a DTensor and it does not hit the previous ValueError?

awgu replied:
We are expecting that `DTensor` has already moved the tensor to the device. If that happened, then `if tensor.device == device: continue` runs. This assertion is just confirming that the `DTensor` behavior is what we assumed.

wanchaol reviewed (approved):
lgtm! have some questions and suggestions inlined.

wanchaol commented:
> @wanchaol mentioned that there are some complications for `(Shard(0), Shard(0))` placement if we shard with TP then FSDP from the state dict perspective. We should flesh this issue out to get everyone on the same page.

Yeah I think because `(Shard(0), Shard(0))` on a 2D mesh happening from left to right from a logical order of sharding prospective, but if we shard TP first then FSDP later it reverses the order, we can solve this via the Strided sharding placement enhancement I have been thinking of.

> If the meta-device module already has DTensors, we expect DTensor's __torch_dispatch__ to handle the initialization ops correctly.

yes the benefits of having DTensor to represent both sharding is that meta device init become uniform too.

> In (3), will FSDP know how to initialize the model, if it came 'meta initialized'? 

@wconstab I think particularly for this PR it does not seem to handle the meta init logic, and I think we should handle the meta init outside of parallelisms in a separate API as @awgu mentioned, so that everything uniformly applied with meta initialization and user don't have to materialize different tensors separately.

wanchaol commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -50,6 +57,79 @@ def test_normalize_device_cuda(self):
+                self.assertEqual(tensor._local_tensor.device, torch.device("cpu"))
+        regex = r"Requires DTensor to have mesh of the same type as the FSDP mesh but got cpu for DTensor and cuda for FSDP"
+        with self.assertRaisesRegex(ValueError, regex):
+            fully_shard(model, mesh=dp_mesh)
```
nit: do you want to add a testing errorness when the model have meta device?

wconstab commented:
ok, lets assume we stuck with PP, TP, DP ordering.

agreed this PR doesn't have to do meta init literally. (we can add later)

for device arg, remind me again why we need that- if users pass a model into fsdp, do they pass it on CPU bc it's too big for gpu?  then we need a device arg to handle moving to gpu?  if we're starting over, can we just say no more cpu -> gpu for fsdp, if your model is big use meta device?  Then FSDP's logic could be simpler-  If the model is meta, register a hook on wrapped parameters so we can observe .to later on.  If the model is cpu/gpu, just assume that's where the user is running it.

awgu commented:
> for device arg, remind me again why we need that- if users pass a model into fsdp, do they pass it on CPU bc it's too big for gpu? then we need a device arg to handle moving to gpu? if we're starting over, can we just say no more cpu -> gpu for fsdp, if your model is big use meta device? Then FSDP's logic could be simpler- If the model is meta, register a hook on wrapped parameters so we can observe .to later on. If the model is cpu/gpu, just assume that's where the user is running it.

This is a good point. Let me share some thoughts for more discussion:
- For `DTensor` APIs, constructing a `DTensor` from a CPU local tensor on a CUDA `DeviceMesh` results in a `DTensor` on the CUDA device.
- Per-parameter FSDP is like a higher-level extension to `DTensor` that manually handles the all-gather/reduce-scatter, bucketing them for performance (along with some other functionality like resharding after forward and mixed precision). It would match `DTensor`'s semantics if `fully_shard` similarly moved CPU parameters/buffers to CUDA if given a CUDA `DeviceMesh`.
- What if we discard the `device` argument and instead rely on (1) `mesh.device_type` and (2) the current device (e.g. `torch.device()` context), where `mesh.device_type`.
    - For meta device, we cannot only rely on `mesh.device_type` since there is no such `meta`-type `DeviceMesh`. Instead, we should only use meta device if under a `torch.device("meta")` context. If under `torch.device("meta")`, then we keep parameters/buffers on meta.
        - I am not sure how to tell what the current device is -- one way is to just do `torch.empty(0)` and see what device it has, so this should not be a blocker.
    - For CUDA device, we use `torch.cuda.current_device()`. If the user wants to use a specific device, it should be set via either `torch.device()` or `torch.cuda.set_device()`.
    - For CPU device, we only use it if `mesh.device_type == "cpu"`.
    - This seems consistent with the `DTensor` behavior.

Let me know what you guys think about this.

Commit:
Update on "[FSDP2] Moved managed states to `device`"


This PR adds logic to move managed states (i.e. parameters/buffers) to `device` (where `device` is the arg passed to `fully_shard`).
- In the typical case, `device` is a CUDA device.
- Similar to `DTensor`'s constructor, `fully_shard` ensures that the managed states are on the `fully_shard` `mesh`'s device before the `fully_shard` call returns.
- We justify this behavior with a few reasons:
    - This is similar to `DTensor`'s constructor, which moves the tensor to the mesh's device.
    - Since FSDP training will use `device` anyway, moving states to `device` makes sharding and hence initialization faster, and it obviates the need for an extra `module.cuda()` (or similar) call after applying FSDP.
    - In the short term, wrapper subclasses like `DTensor` do not support `nn.Module.to` since they do not implement `.data` assignment correctly. The fix to use `torch.utils.swap_tensors` is coming but has not yet landed. Having FSDP ensure that states are already on `device` bypasses `nn.Module.to` not working.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
> For DTensor APIs, constructing a DTensor from a CPU local tensor on a CUDA DeviceMesh results in a DTensor on the CUDA device.

well i might also ask the same questions about DTensor, could we simplify it and use meta or explicit cuda from user?  no more automatic cpu->cuda anywhere?

> What if we discard the device argument and instead rely on (1) mesh.device_type and (2) the current device (e.g. torch.device() context).

What about taking signal from the .parameters on the module you're wrapping?  if they are meta, that tells us we're doing the meta path, and we don't need to assert devicemesh matches until later when we init.  If they are cuda/cpu, could we simply assert the devicemesh matches?

awgu commented:
> What about taking signal from the .parameters on the module you're wrapping? if they are meta, that tells us we're doing the meta path, and we don't need to assert devicemesh matches until later when we init. If they are cuda/cpu, could we simply assert the devicemesh matches?

What if the `.parameters()` do not all have the same device? For example, a user may manually choose to initialize a parameter/buffer not on meta but use meta for everything else. How should FSDP infer the device in that case? Are we relying on the ordering of `"meta"` < `"cpu"` < `"cuda"` and taking the 'largest' device?

awgu commented:
> well i might also ask the same questions about DTensor, could we simplify it and use meta or explicit cuda from user? no more automatic cpu->cuda anywhere?

This puts a large burden on our meta-device init being complete and robust. For the vast majority of cases, FSDP wants the modules on CUDA before training starts; however, for many of those cases, the modules cannot start on CUDA before passing to FSDP. The module can only be on CUDA when passed to FSDP when doing in-lined manual wrapping like xlformers.

Given that we do not have an explicit design for the meta-device init, I want to be wary of introducing this dependency that FSDP cannot be used in many existing cases without this.

> well i might also ask the same questions about DTensor, could we simplify it and use meta or explicit cuda from user? no more automatic cpu->cuda anywhere?

This seems like a UX vs. complexity tradeoff.

The way I am thinking about it are along the lines of:
- What would a failure case look like if we do not auto-move to CUDA?
    - The error would probably be in the 1st collective or compute op, complaining of a device mismatch.
    - For FSDP, we could add some check at lazy initialization time to ensure that managed parameters/buffers were moved to the correct device.
    - For `DTensor`, would we have to have this check before every op? That may be unnecessary CPU overhead for the all but the 1st op.
- When would the user not want us to auto-move to the device as described in the above proposal?
    - I am not sure. If there is not a good reason, then it seems okay for FSDP and `DTensor` to include some (relatively simple) logic to move tensors to device for the user.

Overall, my personal feeling is that the hit against the UX might be greater than the complexity saved.

wanchaol commented:
> This puts a large burden on our meta-device init being complete and robust. For the vast majority of cases, FSDP wants the modules on CUDA before training starts; however, for many of those cases, the modules cannot start on CUDA before passing to FSDP. The module can only be on CUDA when passed to FSDP when doing in-lined manual wrapping like xlformers.

I agree with this. I think our meta init story is still very much early and requires many manual work from the user to init the model. for models that could fit into CPU but not GPU, allowing user to initialize the model on CPU and then pass to our distributed API to do the sharding them move to GPU is still better UX compare to user have to specify how to initialize individual modules/paramters for a model with many layers (the former requires no user code).

That being said, I think if we had a robust meta device init story already without user doing much work, then we should go for only between meta or GPU (without CPU to be a temp ground). Since we do not yet, I would prefer we make both approach available and CPU init workflow can at least work.

> What if we discard the device argument and instead rely on (1) mesh.device_type and (2) the current device (e.g. torch.device() context)

I think we can try to standardize on this to make both FSDP and DTensor API behavior in alignment, I think majority of the time user already set a device via either `torch.cuda.set_device` or `CUDA_VISIBLE_DEVICE` after initialize the process group, and if they are using `device_mesh` instead of process group, currently device mesh would [by default help user to select the device on each process](https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L246)

> For meta device, we cannot only rely on mesh.device_type since there is no such meta-type DeviceMesh. Instead, we should only use meta device if under a torch.device("meta") context. If under torch.device("meta"), then we keep parameters/buffers on meta.

My understanding is that the module parameters/buffers are on meta device, however the device mesh we initialized or user initialized is still on the "target" device, either "cuda" or other backend, so we initialize the model with `torch.device("meta")` but the device mesh is still on CUDA and with torch.cuda.set_device already happened. So when "materializing" the meta tensors, we can simply use `torch.cuda.current_device`

awgu commented:
I squashed this PR into a preceding one to save CI cost.
